src/base/QXmppHttpUploadIq.h

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (C) 2008-2019 The QXmpp developers
+ *
+ * Authors:
+ *  Linus Jahn <lnj@kaidan.im>
+ *
+ * Source:
+ *  https://github.com/qxmpp-project/qxmpp
+ *
+ * This file is a part of QXmpp library.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ */
+
+#ifndef QXMPPHTTPUPLOADIQ_H
+#define QXMPPHTTPUPLOADIQ_H
+
+#include <QMap>
+#include <QMimeType>
+#include <QUrl>
+
+#include "QXmppIq.h"
+
+/// \brief Represents an HTTP File Upload IQ for requesting an upload slot as
+/// defined by XEP-0363: HTTP File Upload [v0.9.0].
+///
+/// \ingroup Stanzas
+
+class QXMPP_EXPORT QXmppHttpUploadRequestIq : public QXmppIq
+{
+public:
+    QString fileName() const;
+    void setFileName(const QString &filename);
+
+    qint64 size() const;
+    void setSize(const qint64 &size);
+
+    QMimeType contentType() const;
+    void setContentType(const QMimeType &type);
+
+    static bool isHttpUploadRequestIq(const QDomElement &element);
+
+protected:
+    /// \cond
+    void parseElementFromChild(const QDomElement &element);
+    void toXmlElementFromChild(QXmlStreamWriter *writer) const;
+    /// \endcond
+
+private:
+    QString m_fileName;
```

Comment:
I tried to put members into an opaque structure wherever possible to preserve ABI in case of future changes, does this make sense here too?

Reply:
I think we should have a private class, yes. I didn't expect roster IQs to change either, but now it turns out your other PR is going to break ABI..

Reply:
Then we should also use a d-pointer for `QXmppMixIq` (I could imagine changes there as the protocol isn't finished yet and hasn't been practically used).